astroid/nodes/_base_nodes.py

```diff
@@ -220,3 +247,524 @@ def _elsed_block_range(
+                if isinstance(owner, (nodes.ClassDef, bases.Instance)):
+                    frame = (
+                        owner if isinstance(owner, nodes.ClassDef) else owner._proxied
+                    )
+                    context.constraints[self.attrname] = get_constraints(
+                        self, frame=frame
+                    )
+                yield from owner.igetattr(self.attrname, context)
+            except (
+                AttributeInferenceError,
+                InferenceError,
+                AttributeError,
+            ):
+                pass
+            finally:
+                context.boundnode = old_boundnode
+        return InferenceErrorInfo(node=self, context=context)
+
+
+class AssignNode(NodeNG):
+    @decorators.raise_if_nothing_inferred
+    @decorators.path_wrapper
+    def _infer_assign(
+        self,
+        context: InferenceContext | None = None,
+        **kwargs: Any,
+    ) -> Generator[InferenceResult, None, None]:
+        """Infer a AssignName/AssignAttr: need to inspect the RHS part of the
+        assign node.
+        """
+        if isinstance(self.parent, nodes.AugAssign):
+            return self.parent.infer(context)
+
+        stmts = list(self.assigned_stmts(context=context))
+        return bases._infer_stmts(stmts, context)
+
+    def assigned_stmts(
+        self: nodes.AssignName | nodes.AssignAttr,
+        node: AssignedStmtsPossibleNode = None,
+        context: InferenceContext | None = None,
+        assign_path: list[int] | None = None,
+    ) -> Any:
+        """Returns the assigned statement (non inferred) according to the assignment type."""
+        return self.parent.assigned_stmts(node=self, context=context)
+
+
+class NameNode(LookupMixIn):
+    name: str
+
+    @decorators.raise_if_nothing_inferred
+    def _infer_name_node(
+        self,
+        context: InferenceContext | None = None,
+        **kwargs: Any,
+    ) -> Generator[InferenceResult, None, None]:
+        """Infer a Name: use name lookup rules."""
+        # pylint: disable=import-outside-toplevel
+        from astroid.constraint import get_constraints
+        from astroid.helpers import _higher_function_scope
+
+        frame, stmts = self.lookup(self.name)
+        if not stmts:
+            # Try to see if the name is enclosed in a nested function
+            # and use the higher (first function) scope for searching.
+            parent_function = _higher_function_scope(self.scope())
+            if parent_function:
+                _, stmts = parent_function.lookup(self.name)
+
+            if not stmts:
+                raise NameInferenceError(
+                    name=self.name, scope=self.scope(), context=context
+                )
+        context = copy_context(context)
+        context.lookupname = self.name
+        context.constraints[self.name] = get_constraints(self, frame)
+
+        return bases._infer_stmts(stmts, context, frame)
```

Comment:
I really like these new base classes, but don't really like how long this file is going to be and how many different types of bases nodes are in here. Not sure if this is fixable though.

Reply:
We could burst some files later on but we have other issues to deal with imo. (Especially cyclic imports that make this harder)